ICRCs/ICRC-123/ICRC-123.md

```diff
@@ -0,0 +1,223 @@
+
+### Ledger Enforcement Rules
+
+- **Transfers:**
+    - A ledger **MUST reject** any `icrc1_transfer` or `icrc2_transfer_from` transaction where the **sender** account (the `from` account in the operation) is currently RESTRICTED.
+    - The ledger **MAY**, according to its policy, also reject `icrc1_transfer` or `icrc2_transfer_from` transactions if the **recipient** account (the `to` account in the operation) is RESTRICTED, or it MAY allow incoming funds to a RESTRICTED recipient.
+- **ICRC-2 Operations:**
+    - **`icrc2_approve` (Granting Approval):** If an account is RESTRICTED, its owner **MUST NOT** be able to authorize an `icrc2_approve` transaction where this restricted account is the one granting the approval (i.e., the `account` argument in `icrc2_approve` which specifies the owner of the funds being approved for spending).
+    - **`icrc2_approve` (Receiving Approval):** The ledger's policy SHOULD define whether an approval can be granted *to* a RESTRICTED account (i.e., a RESTRICTED account being the `spender` argument in an `icrc2_approve` call initiated by an unrestricted account owner). Even if an approval is granted to a RESTRICTED account, that account **MUST NOT** be able to use this approval (e.g., by calling `icrc2_transfer_from`) while it remains RESTRICTED.
+    - **`icrc2_transfer_from` (Acting as Spender):** An account that is RESTRICTED **MUST NOT** be able to initiate an `icrc2_transfer_from` call (i.e., act as an approved spender), even if it holds a valid approval for another account.
+- Freeze and unfreeze blocks do **not** modify or invalidate previous transactions. They apply only to transactions attempted **at or after** the block height at which the freeze/unfreeze block is recorded and its state change takes effect.
+- Freeze and unfreeze blocks MUST be **permanently recorded** and included in the block hash chain.
+
+### Idempotency and Redundancy
+
+- A ledger MAY reject freeze or unfreeze blocks that would have **no effect** on the current RESTRICTED status of the target account or principal (e.g., freezing an already frozen account via the same mechanism), or MAY choose to **record them anyway** for auditability.
+- Clients interpreting freeze status MUST follow the **"latest-action-wins" rule** as defined in "Account Status": the most recent freeze or unfreeze block affecting an account or principal determines its effective status.
+
+### Querying Freeze Status
+
+Ledgers implementing this standard SHOULD expose a query interface (e.g., `is_account_restricted(account): bool`) that returns whether an account is currently RESTRICTED according to the rules defined in "Account Status". This serves as a convenience layer and does not replace auditing based on block history.
```

Comment:
shouldn't we clearly define the query interface to be used? what's the benefit of leaving this up to the implementer?

I think the ecosystem would benefit of having query interfaces clearly defined.

Reply:
I mean we could leave it up to the implementer to decide whether to expose such function or not. but if they decide to implement it, it would be good to have a common/reusable function name for that.

clients might need to implement some error handling though if this function is not present.